manifests/0000_90_kube-controller-manager-operator_05_alerts.yaml

```diff
@@ -39,3 +39,27 @@ spec:
           for: 15m
           labels:
             severity: critical
+        - alert: GarbageCollectorSyncFailed
+          annotations:
+            summary: There was a problem with syncing the resources for garbage collection.
+            description: Garbage Collector had a problem with syncing and monitoring the available resources. Please see KubeControllerManager logs for more details.
+          expr: |
+            rate(garbagecollector_controller_garbagecollector_resources_sync_error_total{}[1h]) > 0
+          labels:
+            severity: warning
+        - alert: GarbageCollectorObjectDeletionFailed
+          annotations:
+            summary: There was a problem with deleting an object.
+            description: Garbage Collector had some problems when attempting to delete an object. Please see KubeControllerManager logs for more details.
+          expr: |
+            rate(workqueue_retries_total{name="garbage_collector_attempt_to_delete"}[5m]) > 0
+          labels:
+            severity: warning
+        - alert: GarbageCollectorObjectOrphaningFailed
+          annotations:
+            summary: There was a problem with orphaning an object.
+            description: Garbage Collector had a problem when attempting to orphan an object. Please see KubeControllerManager logs for more details.
+          expr: |
+            rate(workqueue_retries_total{name="garbage_collector_attempt_to_orphan"}[5m]) > 0
```

Comment:
I changed the configuration of these alerts (from `24h` to `5m`) but there are couple of problems with GC queues

1. It can have transitory errors, that are just for the purpose of retring the queue (https://github.com/kubernetes/kubernetes/blob/dd1a7893d4de47cf75cd6a11f38474b318dcc9b2/pkg/controller/garbagecollector/garbagecollector.go#L372). Since that can happen quite often we cannot observe the error rate change over long time. I have put 5 minutes there, so the alert will stop firing after that. Example errors from CI logs (they happen quite often in each run):


```
2022-05-11T21:42:35.109736421Z E0511 21:42:35.109708       1 garbagecollector.go:347] error syncing item &garbagecollector.node{identity:garbagecollector.objectReference{OwnerReference:v1.OwnerReference{APIVersion:"discovery.k8s.io/v1", Kind:"EndpointSlice", Name:"example-custom-endpoints-ff6nf", UID:"faae0457-1778-46d7-9cd1-6278ea6778f1", Controller:(*bool)(nil), BlockOwnerDeletion:(*bool)(nil)}, Namespace:"e2e-endpointslicemirroring-8071"}, dependentsLock:sync.RWMutex{w:sync.Mutex{state:0, sema:0x0}, writerSem:0x0, readerSem:0x0, readerCount:1, readerWait:0}, dependents:map[*garbagecollector.node]struct {}{}, deletingDependents:false, deletingDependentsLock:sync.RWMutex{w:sync.Mutex{state:0, sema:0x0}, writerSem:0x0, readerSem:0x0, readerCount:0, readerWait:0}, beingDeleted:false, beingDeletedLock:sync.RWMutex{w:sync.Mutex{state:0, sema:0x0}, writerSem:0x0, readerSem:0x0, readerCount:0, readerWait:0}, virtual:false, virtualLock:sync.RWMutex{w:sync.Mutex{state:0, sema:0x0}, writerSem:0x0, readerSem:0x0, readerCount:0, readerWait:0}, owners:[]v1.OwnerReference{v1.OwnerReference{APIVersion:"v1", Kind:"Endpoints", Name:"example-custom-endpoints", UID:"dc339dea-24f0-4a6f-a1f8-64b5d27e3ff6", Controller:(*bool)(0xc00a68b74c), BlockOwnerDeletion:(*bool)(0xc00a68b74d)}}}: endpointslices.discovery.k8s.io "example-custom-endpoints-ff6nf" not found
2022-05-11T21:42:35.116620452Z I0511 21:42:35.116595       1 garbagecollector.go:468] "Processing object" object="e2e-endpointslicemirroring-8071/example-custom-endpoints-ff6nf" objectUID=faae0457-1778-46d7-9cd1-6278ea6778f1 kind="EndpointSlice" virtual=false
2022-05-11T21:42:35.118599158Z I0511 21:42:35.116762       1 garbagecollector.go:580] "Deleting object" object="e2e-container-probe-3234/deployer-dockercfg-cqqnt" objectUID=94979b07-d97f-48c5-9a85-e5ac3fa71bdb kind="Secret" propagationPolicy=Background
2022-05-11T21:42:35.118599158Z E0511 21:42:35.116803       1 garbagecollector.go:347] error syncing item &garbagecollector.node{identity:garbagecollector.objectReference{OwnerReference:v1.OwnerReference{APIVersion:"v1", Kind:"Secret", Name:"builder-dockercfg-mm68z", UID:"7bf8d21e-79f5-4227-a0bd-bec4a113d826", Controller:(*bool)(nil), BlockOwnerDeletion:(*bool)(nil)}, Namespace:"e2e-container-probe-3234"}, dependentsLock:sync.RWMutex{w:sync.Mutex{state:0, sema:0x0}, writerSem:0x0, readerSem:0x0, readerCount:1, readerWait:0}, dependents:map[*garbagecollector.node]struct {}{}, deletingDependents:false, deletingDependentsLock:sync.RWMutex{w:sync.Mutex{state:0, sema:0x0}, writerSem:0x0, readerSem:0x0, readerCount:0, readerWait:0}, beingDeleted:false, beingDeletedLock:sync.RWMutex{w:sync.Mutex{state:0, sema:0x0}, writerSem:0x0, readerSem:0x0, readerCount:0, readerWait:0}, virtual:false, virtualLock:sync.RWMutex{w:sync.Mutex{state:0, sema:0x0}, writerSem:0x0, readerSem:0x0, readerCount:0, readerWait:0}, owners:[]v1.OwnerReference{v1.OwnerReference{APIVersion:"v1", Kind:"Secret", Name:"builder-token-x4t5h", UID:"3fe0d74b-cc1e-437f-8354-944c091a4646", Controller:(*bool)(0xc00966a2a8), BlockOwnerDeletion:(*bool)(0xc00966a2a9)}}}: secrets "builder-dockercfg-mm68z" not found
2022-05-11T21:42:35.123943615Z I0511 21:42:35.122021       1 garbagecollector.go:468] "Processing object" object="e2e-container-probe-3234/builder-dockercfg-mm68z" objectUID=7bf8d21e-79f5-4227-a0bd-bec4a113d826 kind="Secret" virtual=false2022-05-11T21:42:35.151610596Z E0511 21:42:35.151577       1 garbagecollector.go:347] error syncing item &garbagecollector.node{identity:garbagecollector.objectReference{OwnerReference:v1.OwnerReference{APIVersion:"v1", Kind:"Secret", Name:"deployer-dockercfg-cqqnt", UID:"94979b07-d97f-48c5-9a85-e5ac3fa71bdb", Controller:(*bool)(nil), BlockOwnerDeletion:(*bool)(nil)}, Namespace:"e2e-container-probe-3234"}, dependentsLock:sync.RWMutex{w:sync.Mutex{state:0, sema:0x0}, writerSem:0x0, readerSem:0x0, readerCount:1, readerWait:0}, dependents:map[*garbagecollector.node]struct {}{}, deletingDependents:false, deletingDependentsLock:sync.RWMutex{w:sync.Mutex{state:0, sema:0x0}, writerSem:0x0, readerSem:0x0, readerCount:0, readerWait:0}, beingDeleted:false, beingDeletedLock:sync.RWMutex{w:sync.Mutex{state:0, sema:0x0}, writerSem:0x0, readerSem:0x0, readerCount:0, readerWait:0}, virtual:false, virtualLock:sync.RWMutex{w:sync.Mutex{state:0, sema:0x0}, writerSem:0x0, readerSem:0x0, readerCount:0, readerWait:0}, owners:[]v1.OwnerReference{v1.OwnerReference{APIVersion:"v1", Kind:"Secret", Name:"deployer-token-mqt88", UID:"af5e1d10-e2be-4ae3-bf3a-f30c44e9bbb9", Controller:(*bool)(0xc00a711b28), BlockOwnerDeletion:(*bool)(0xc00a711b29)}}}: secrets "deployer-dockercfg-cqqnt" not found
2022-05-11T21:42:35.157711325Z I0511 21:42:35.156666       1 garbagecollector.go:468] "Processing object" object="e2e-container-probe-3234/deployer-dockercfg-cqqnt" objectUID=94979b07-d97f-48c5-9a85-e5ac3fa71bdb kind="Secret" virtual=false
```


2. The negative of the 5 minutes observation period is that the alert will miss long term errors since we have exponential backoff and no periodic resync in GC (as far as I could see).


So basically the usefullness of queue alerts is limited and could be useful only in a case something goes very wrong and we have a lot of new errors appearing very often (eg 1 a minute)

Reply:
having before dismissed GC debug handler would help here a lot (https://github.com/kubernetes/kubernetes/pull/105705). We could easily distinguish between transient errors (simple queue requeues) and permanent errors by knowing the current state.

With current metrics
- we need a short observation period for the transient errors to quickly dismiss them (eg long alerts will botch CI)
- but we need long observation periods to guess the exponential backoff of the persistent errors

Which is hard to do without knowing the cause of each error. Even if we expose that somehow via some new metrics (which I think is impossilble without adding non trivial logic to gc) it is not possible to deal with exponential backoff easily - even if we can have 24h observation period it is not the best solution as the alert will stop if nobody gives it attention for some time

These limitation were not clear to me when we discussed it last time.

The other alternative is just checking the dependencies of GC like APIServices, as we also discussed. But I am not favoring this approach since we would not have a real insight into what GC is doing wrt errors.

@soltysh thoughts?

Reply:
@atiratree let's leave that APIServices as a last resort, for now is it possible to pair this metric with some others related with workqueue, `workqueue_depth`, for example. In that case transient errors will grow the depth, but will eventually stop, whereas in the long observation both can potentially grow. Or maybe a different combination, worth syncing with @dgrisonnet about that it.

Reply:
I think for the alert to be at warning level, we decided to use 60 min as timeline.

https://github.com/openshift/enhancements/blob/master/enhancements/monitoring/alerting-consistency.md#warning-alerts

> Even if we expose that somehow via some new metrics (which I think is impossilble without adding non trivial logic to gc) it is not possible to deal with exponential backoff easily

Would it be easier to split the metric and make it much more specialized to indicate that the discovery of objects is failing(perhaps a failing webhook conversion) by having a metric [here](https://github.com/atiratree/kubernetes/blob/2fc401d0a295dd30ec73e99dbcdbba6a08268d6e/pkg/controller/garbagecollector/garbagecollector.go#L763)

Reply:
here the period has a bit different meaning, since we use it just for stopping caring about the error - and we for sure know that the for example sync error has happened. 

> Would it be easier to split the metric and make it much more specialized to indicate that the discovery of objects is failing(perhaps a failing webhook conversion) by having a metric

I am not sure if another metric in this place would be beneficial yet

Reply:
I went over all available metrics and there seems not to be a solution:

`workqueue_retries_total`: does not work due to the already mentioned reasons
`workqueue_depth` and `workqueue_queue_duration_seconds`: does not work - since retried items are not really in the queue and thus are not registred via the metric and are just waiting to be added to the queue after a backoff
`workqueue_work_duration_seconds`: not usable - tracks only work duration
`workqueue_adds_total`: not usable - tracks only adds
`workqueue_unfinished_work_seconds`: not usable (usually 0)
`workqueue_longest_running_processor_seconds`: not usable (usually 0)

So wrt alternatives:

It cannot be simply solved just by looking at APIServices to detect they are down. We might still have old items in GC (errors / retries) even after the APIServices get removed,


To reiterate, the problem is that the same rest mapping errors can be both transient and persistent (https://github.com/kubernetes/kubernetes/blob/dd1a7893d4de47cf75cd6a11f38474b318dcc9b2/pkg/controller/garbagecollector/garbagecollector.go#L372-L374) as this depends on an outside actor/sync.
I think it could be solved generically by a mechanism that would track all items that are waiting for a requeue/retry and report a metric if they overcome an arbitrary threshold (eg. 30s) to inhibit counting of legitimate retries.

This would require some changes in a GC to wire this. @soltysh @ravisantoshgudimetla  thoughts?

Reply:
I am not sure I have a complete understanding of the problem you are trying to solve, but I am happy to discuss it offline if you want.

Reply:
After discussing this with @ravisantoshgudimetla and @dgrisonnet, the best course is to introduce new metrics in GC for the delete and orphan queue (probably a histogram for the backoffs).